support/node-health-check/src/main.rs

```diff
@@ -0,0 +1,107 @@
+use std::convert::Infallible;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::{sync::Arc, time::Duration};
+
+use anyhow::Result;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use structopt::StructOpt;
+use warp::Filter;
+
+/// Health struct returned by the RPC
+#[derive(Debug, PartialEq, Serialize, Deserialize, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Health {
+    /// Number of connected peers
+    pub peers: usize,
+    /// Is the node syncing
+    pub is_syncing: bool,
+    /// Should this node have any peers
+    /// Might be false for local chains or when running without discovery.
+    pub should_have_peers: bool,
+}
+
+#[derive(Debug, StructOpt, Clone)]
+#[structopt(
+    name = "node-health-check",
+    about = "Sidecar for substrate node to report health compatible with AWS ALB."
+)]
+struct Opt {
+    /// Debug
+    #[structopt(short, long)]
+    debug: bool,
+
+    /// Set poll_interval
+    #[structopt(short = "i", default_value = "5")]
+    poll_interval: u64,
+
+    /// Set minimum peers
+    #[structopt(default_value = "2")]
+    min_peers: usize,
+
+    /// Node rpc endpoint
+    #[structopt(short, long, env, default_value = "http://localhost:9933/health")]
+    node_rpc_endpoint: String,
+
+    // The port to listen on
+    #[structopt(short, long, default_value = "9955")]
+    port: u16,
+}
+
+async fn poller(opts: Opt, healthy: Arc<AtomicBool>) {
+    loop {
+        let resp = async {
+            reqwest::get(&opts.node_rpc_endpoint)
+                .await?
+                .json::<Health>()
+                .await
+        };
+
+        match resp.await {
+            Ok(health) => {
+                let node_is_healthy = !health.is_syncing
+                    && (!health.should_have_peers || health.peers >= opts.min_peers);
+
+                if opts.debug {
+                    eprintln!("Response = {:?}", health);
+                    eprintln!("Healthy  = {:?}", node_is_healthy);
+                }
+                healthy.store(node_is_healthy, Ordering::Relaxed);
+            }
+            Err(error) => {
+                eprintln!("{:?}", error);
+                healthy.store(false, Ordering::Relaxed);
+            }
+        }
+
+        tokio::time::sleep(Duration::from_secs(opts.poll_interval)).await;
+    }
+}
+
+async fn return_health_status(healthy: Arc<AtomicBool>) -> Result<impl warp::Reply, Infallible> {
+    if healthy.load(Ordering::Relaxed) {
+        Ok(StatusCode::OK)
+    } else {
+        Ok(StatusCode::NOT_FOUND)
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let opts = Opt::from_args();
+    let healthy = Arc::new(AtomicBool::new(true));
+
+    let health_check = {
+        let healthy = healthy.clone();
+        warp::path!("health")
+            .map(move || healthy.clone())
```

Comment:
the thing is, warp executes the move closure on each request, and delivers is to the `and_then` function as a parameter. This is why there has to be clone there, as is will create a new ref on each request.

Reply:
If you have return health status take a `&Arc`, or possibly even `&AtomicBool` then it can avoid the clone. I'd think this would work:
```rust
{
  let healthy = healthy.clone();
  warp::path!("health").and_then(|_| return_health_status(&healthy))
}
```

Reply:
It would be great if that would work, but it does not. I get this message: 
![image](https://user-images.githubusercontent.com/300333/135843615-a092b6c7-153e-4736-925f-93cc75790376.png)

Reply:
and when I use the move it again complaints about the lifetimes:

![image](https://user-images.githubusercontent.com/300333/135843955-737bb36d-f9d6-47d4-8f23-32c6bb83d156.png)